File: python_mid_june_2024/15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        def twoSum(idx: int, target: int):
            i, j = idx+1, n-1
            
            while i < j:
                total = nums[i] + nums[j]
                if total == target:
                    res.append([nums[idx], nums[i], nums[j]])
                    i += 1
                    j -= 1
                    while i < j and nums[i] == nums[i-1]:
                        i += 1
                elif total < target:
                    i += 1
                else:
                    j -= 1
        
        nums.sort()
        i, n = 0, len(nums)
        res = []
        for i in range(n):
            if nums[i] > 0:
                break
            if i == 0 or nums[i-1] != nums[i]:
                twoSum(i, -nums[i])
            
        return res
            

File: java_august/15. 3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        List<List<Integer>> answer = new ArrayList<>();
        for (int i=0; i<n && nums[i] <= 0; i++) {
            if (i == 0 || nums[i-1] != nums[i]) {
                twoSum(nums, i, answer);
            }
        }
        return answer;
    }

    public void twoSum(int[] nums, int i, List<List<Integer>> answer) {
        int lo = i+1, hi = nums.length-1;
        while (lo < hi) {
            int sum = nums[i] + nums[lo] + nums[hi];
            if (sum == 0) {
                answer.add(Arrays.asList(nums[i], nums[lo++], nums[hi--]));
                while (lo < hi && nums[lo] == nums[lo-1]) {
                    lo++;
                }
            }
            else if (sum < 0) {
                lo++;
            }
            else{
                hi--;
            }
        }
    }
}